### PR TITLE
feat(core): allow plugins to declare custom route context

### PR DESCRIPTION
**Pull request by Josh-Cena**
## Motivation

Complete the TODO. Honestly I'm not too sure what we should put here, so I didn't modify the content plugins yet. We already have front matter duplicated twice (once in `metadata.frontMatter`, once as separate `frontMatter` injected from MDX loader), I don't really want to duplicate it a third time as that would be disastrous for bundle size.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

Builds same as before

### CODE REVIEW

**Review by slorber — CHANGES_REQUESTED** [review]
LGTM 👍 better impl than my first attempt

File: packages/docusaurus/src/server/plugins/index.ts [hunk 1]
```diff
@@ -104,22 +103,15 @@ export async function loadPlugins(context: LoadContext): Promise<{
         plugin.name,
         pluginId,
       );
-      // TODO this would be better to do all that in the codegen phase
-      // TODO handle context for nested routes
```
[the remark this answers is not in the document]
Comment (Josh-Cena) [reply packages/docusaurus/src/server/plugins/index.ts:108]: It works everywhere, because it's handled in the same way as route components are handled. As long as there are `RouteContextProvider`s, the nested routes will be properly merged. Unless "nested routes" mean something different here?

File: packages/docusaurus/src/server/plugins/index.ts [hunk 2]
```diff
@@ -104,22 +103,15 @@ export async function loadPlugins(context: LoadContext): Promise<{
...
-      };
       const pluginRouteContextModulePath = path.join(
         dataDir,
         `${docuHash('pluginRouteContextModule')}.json`,
       );
       await generate(
         '/',
         pluginRouteContextModulePath,
-        JSON.stringify(pluginRouteContext, null, 2),
+        JSON.stringify({name: plugin.name, id: pluginId}, null, 2),
```
[the remark this answers is not in the document]
Comment (Josh-Cena) [reply packages/docusaurus/src/server/plugins/index.ts:113]: I'm really looking forward to `satisfies` in TS 4.7

File: packages/docusaurus/src/server/plugins/index.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (slorber) [reply packages/docusaurus/src/server/plugins/index.ts:108]: Hmm, I think it "works" yes, but only parent routes are namespacing the context data under a "data" attribute while nested routes expose the context data

Maybe if we had a good case to test this feature on docs route we'll see if it works?

This looks fine for now, we'll fix edge cases as soon as we have a use-case for nested context

File: packages/docusaurus/src/server/routes.ts [hunk 3]
```diff
@@ -272,7 +272,9 @@ ${JSON.stringify(routeConfig)}`,
   res.routesChunkNames[`${routePath}-${routeHash}`] = {
     // Avoid clash with a prop called "component"
     ...genChunkNames({__comp: component}, 'component', component, res),
-    ...genChunkNames({__context: context}, 'context', routePath, res),
+    ...(context
+      ? genChunkNames({__context: context}, 'context', routePath, res)
+      : {}),
```
Comment (slorber) [inline packages/docusaurus/src/server/routes.ts:277]: not important but you could replace ternary by && here

image
